examples/fakerealtime/main.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Ben Darnell
+
+// This example uses a log-style table in an approximation of the
+// "fake real time" system used at Friendfeed. Two tables are used: a
+// `messages` table stores the complete data for all messages
+// organized by channel, and a global `updates` table stores metadata
+// about recently-updated channels.
+//
+// The example currently runs a number of writers, which update both
+// tables transactionally. Future work includes implementing a reader
+// which will effectively tail the `updates` table to learn when it
+// needs to query a channel's messages from the `messages` table.
+//
+// The implementation currently guarantees that `update_ids` are
+// strictly monotonic, which is extremely expensive under contention
+// (two writer threads will fight with each other for 20 seconds or
+// more before making progress).
+//
+// One alternate solution is to relax the monotonicity constraint (by
+// using a timestmp instead of a sequential ID), at the expense of
+// making the reader more complicated (since it would need to re-scan
+// records it had already read to see if anything "in the past"
+// committed since it last read). We would still need to put some sort
+// of bound on the reader's scans; any updates that went longer than
+// this without being committed would go unnoticed.
+//
+// Another alternative is to reduce contention by adding a small
+// random number (in range (0, N)) to the beginning of the `updates`
+// table primary key. This reduces contention but makes the reader do
+// N queries to see all updates. This is effectively what Friendfeed
+// did, since there was one `updates` table per MySQL shard.
+package main
+
+import (
+	"database/sql"
+	"flag"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/cockroach/security"
+	"github.com/cockroachdb/cockroach/security/securitytest"
+	"github.com/cockroachdb/cockroach/server"
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/montanaflynn/stats"
+)
+
+func createTables(db *sql.DB) error {
+	statements := []string{
+		`create database if not exists fakerealtime`,
```

Comment:
Yeah, but why? It's just unnecessary work (especially since I don't have capslock mapped on my keyboard). Consider this a test of our case-insensitive parsing :)